pkg/server/plugin/bundlepublisher/gcpcloudstorage/gcpcloudstorage.go

```diff
@@ -0,0 +1,259 @@
+	configv1.UnsafeConfigServer
+
+	config    *Config
+	configMtx sync.RWMutex
+
+	bundle    *types.Bundle
+	bundleMtx sync.RWMutex
+
+	hooks     pluginHooks
+	gcsClient gcsService
+	log       hclog.Logger
+}
+
+// SetLogger sets a logger in the plugin.
+func (p *Plugin) SetLogger(log hclog.Logger) {
+	p.log = log
+}
+
+// Configure configures the plugin.
+func (p *Plugin) Configure(ctx context.Context, req *configv1.ConfigureRequest) (*configv1.ConfigureResponse, error) {
+	config, err := parseAndValidateConfig(req.HclConfiguration)
+	if err != nil {
+		return nil, err
+	}
+
+	var opts []option.ClientOption
+	if config.ServiceAccountFile != "" {
+		opts = append(opts, option.WithCredentialsFile(config.ServiceAccountFile))
+	}
+
+	gcsClient, err := p.hooks.newGCSClientFunc(ctx, opts...)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to create client: %v", err)
+	}
+	p.gcsClient = gcsClient
+
+	p.setConfig(config)
+	p.setBundle(nil)
+	return &configv1.ConfigureResponse{}, nil
+}
+
+// PublishBundle puts the bundle in the configured GCS bucket and object name.
+func (p *Plugin) PublishBundle(ctx context.Context, req *bundlepublisherv1.PublishBundleRequest) (*bundlepublisherv1.PublishBundleResponse, error) {
+	config, err := p.getConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	if req.Bundle == nil {
+		return nil, status.Error(codes.InvalidArgument, "missing bundle in request")
+	}
+
+	currentBundle := p.getBundle()
+	if proto.Equal(req.Bundle, currentBundle) {
+		// Bundle not changed. No need to publish.
+		return &bundlepublisherv1.PublishBundleResponse{}, nil
+	}
+
+	formatter := bundleformat.NewFormatter(req.Bundle)
+	bundleBytes, err := formatter.Format(config.bundleFormat)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "could not format bundle: %v", err.Error())
+	}
+
+	bucketHandle := p.gcsClient.Bucket(config.BucketName)
+	if bucketHandle == nil { // Purely defensive, the Bucket function implemented in GCS always returns a BucketHandle.
+		return nil, status.Error(codes.Internal, "could not get bucket handle")
+	}
+
+	objectHandle := bucketHandle.Object(config.ObjectName)
+	if objectHandle == nil { // Purely defensive, the Object function implemented in GCS always returns an ObjectHandle.
+		return nil, status.Error(codes.Internal, "could not get object handle")
+	}
+
+	storageWriter := p.hooks.newStorageWriterFunc(ctx, objectHandle)
+	if storageWriter == nil { // Purely defensive, the NewWriter function implemented in GCS always returns a storage writer
+		return nil, status.Error(codes.Internal, "could not initialize storage writer")
+	}
+
+	_, err = storageWriter.Write(bundleBytes)
+	// The number of bytes written can be safely ignored. To determine if an
+	// object was successfully uploaded, we need to look at the error returned
+	// from storageWriter.Close().
+	if err != nil {
+		// Close the storage writer before returning.
+		if closeErr := storageWriter.Close(); err != nil {
```

Comment:
why no to use a defer after writer was created?

Reply:
if the `_, err = storageWriter.Write(bundleBytes)` call returns an error, we want to first call Close() and then return the error returned by Write, so I don't think that using `defer` would help here.